docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+
+This ADR seeks to clarify and prioritize the secret handling approach taken by EdgeX.
+
+EdgeX microservices need a number of secrets to be created and distributed
+in order to create a functional, secure system.
+Among these secrets are:
+
+- Privileged administrator passwords (such as a database superuser password)
+- Service account passwords (e.g. for a database)
+- PKI certificates
+
+There is a lack of consistency on how secrets are created and distributed to EdgeX microservices,
+and when developers need to add new components to the system,
+it is unclear on what the preferred approach should be.
+
+This document assumes a threat model wherein the EdgeX services are sandboxed
+(such as in a snap or a container) and the host system is trusted,
+and all services running in a single snap share a trust boundary.
+
+### Current practices survey
+
+A survey on the existing EdgeX secrets reveals the following appoaches.
+
+A designation of "compliant" means that the current implementation
+is aligned with the recommended practices documented in the next section.
+A designation of "non-compliant" means that the current implementation
+uses an implemention mechanism outside of the recommended practices documented in the next section.
+A "non-compliant" implementation is a candidate for refactoring
+to bring the implementation into conformance with the recommended practices.
+
+- PKI private keys
+  * Docker: PKI generated by standalone utility every boot. Distribution via secrets volume. (Compliant.)
+  * Snaps: PKI generated by standalone utility every boot. Deployed to `$SNAP_DATA/secrets`. (Compliant.)
+
+- Secret store master password
+  * Docker: Distribution via persistent docker volume. (Non-compliant.)
+  * Snaps: Stored in `$SNAP_DATA/config/security-secrets-setup/res`. (Non-compliant.)
+
+- Secret store per-service authentication tokens
+  * Docker: Distribution via secrets volume, generated every boot. (Compliant.)
+  * Snaps: Distribution via secrets volume, generated every boot. (Compliant.)
+
+- Postgres superuser password
+  * Docker: Hard-coded into docker-compose file, checked in to source control. (Non-compliant.)
+  * Snaps: Deployed to `$SNAP_DATA/config/postgres/kongpw`. (Non-compliant.) 
+
+- MongoDB service account passwords
+  * Docker: Direct consumption from secret store. (Compliant.)
+  * Snaps: Direct consumption from secret store. (Compliant.)
+
+- Redis authentication password
+  * Docker: Server--staged to secrets volume and injected via command line. (Non-compliant.). Clients--direct consumption from secret store. (Compliant.)
+  * Snaps: Server--staged to `$SNAP_DATA/secrets/edgex-redis/redis5-password` and injected via command line. (Non-compliant.). Clients--direct consumption from secret store. (Compliant.)
+
+- Kong authentication tokens
+  * Docker: System of reference is unencrypted Postgres database. (Non-compliant.)
+  * Snaps: System of reference is unencrypted Postgres database. (Non-compliant.)
```

Comment:
IMO landing them on the file system is no worse than where they are now: unencrypted in a local data file, either binary format (database) or text format (config file)  A config file might even be preferable to returning it over a network channel from a database.  It might even be possible to generate the config file using data from the secret store, which would be a similar approach to passing the Redis password.